src/compiler/checker.ts

```diff
@@ -8431,6 +8434,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             }
 
             function visitExistingNodeTreeSymbolsWorker(node: Node): Node | undefined {
+                if (isJSDocTypeExpression(node)) {

```

Comment:
The constraint type node for a type parameter declared in jsdoc is a `JSDocTypeExpression` which, as it turns out, is not *actually* a type node (according to `isTypeNode`, though the typesystem thinks it is - inconsistent much?), so it needs to get unwrapped; you also definitely don't want the `{}` from jsdoc copied into a `.d.ts` file in the first place! So the assertion triggering off the core of this change found this bug, which was nice.